src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+        "couldn't find '/' in `" + restriction_opt + "'",
+        "--" RESTRICT_FUNCTION_POINTER_OPT,
+        restriction_format_message};
+    }
+    if(pointer_name_end == restriction_opt.size())
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "couldn't find names of targets after '/' in `" + restriction_opt + "'",
+        "--" RESTRICT_FUNCTION_POINTER_OPT,
+        restriction_format_message};
+    }
+    if(pointer_name_end == 0)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "couldn't find target name before '/' in `" + restriction_opt + "'",
+        "--" RESTRICT_FUNCTION_POINTER_OPT};
+    }
+    auto const pointer_name = restriction_opt.substr(0, pointer_name_end);
+    auto const target_names_substring =
+      restriction_opt.substr(pointer_name_end + 1);
+    auto const target_name_strings = split_string(target_names_substring, ',');
+    if(target_name_strings.size() == 1 && target_name_strings[0].empty())
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "missing target list for function pointer restriction " + pointer_name,
+        "--" RESTRICT_FUNCTION_POINTER_OPT,
+        restriction_format_message};
+    }
+    auto const target_names = ([&target_name_strings] {
+      auto target_names = std::unordered_set<irep_idt>{};
+      std::transform(
+        target_name_strings.begin(),
+        target_name_strings.end(),
+        std::inserter(target_names, target_names.end()),
+        string2id);
+      return target_names;
+    })();
+    for(auto const &target_name : target_names)
+    {
+      if(target_name == ID_empty_string)
+      {
+        throw invalid_command_line_argument_exceptiont(
+          "leading or trailing comma in restrictions for `" + pointer_name +
+            "'",
+          "--" RESTRICT_FUNCTION_POINTER_OPT,
+          restriction_format_message);
+      }
+    }
+    if(!function_pointer_restrictions
+          .emplace(irep_idt{pointer_name}, target_names)
+          .second)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "function pointer restriction for `" + pointer_name +
+          "' was specified twice",
+        "--" RESTRICT_FUNCTION_POINTER_OPT};
+    };
+  }
+  return function_pointer_restrictions;
+}
+
+function_pointer_restrictionst::restrictionst
+parse_function_pointer_restrictions_from_file(
+  const std::list<std::string> &filenames,
+  message_handlert &message_handler)
+{
+  auto merged_restrictions = function_pointer_restrictionst::restrictionst{};
+  for(auto const &filename : filenames)
+  {
+    auto const restrictions =
+      function_pointer_restrictionst::read_from_file(filename, message_handler);
+    merged_restrictions = merge_function_pointer_restrictions(
+      merged_restrictions, restrictions.restrictions);
+  }
+  return merged_restrictions;
+}
+} // namespace
+
+function_pointer_restrictionst function_pointer_restrictionst::from_options(
```

Comment:
:bulb: A great candidate for unit tests (which would document expected format for free)

Reply:
Will defer this to the next PR which changes `from_options()`.